### ISSUE TITLE
"CS0675 Bitwise-or operator used on a sign-extended operand" reported too often

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#4027]
See http://stackoverflow.com/questions/31542256/bitwise-or-operator-used-on-a-sign-extended-operand-in-visual-studio-2015, which reports:

> I just tried installing Visual Studio 2015, and when trying to compile an old project, I got the warning
```none
CS0675 Bitwise-or operator used on a sign-extended operand; consider casting to a smaller unsigned type first
```
for a piece of code that does not give the same warning when compiling in Visual Studio 2013. I found out that all it takes to reproduce is this very simple code:
```cs
short a = 0;
int b = 0;
a |= (short)b;
```
Now, I have read this SO question, I have read Eric Lippert's blog post on this issue, and I quickly read up on sign extension, but my understanding is that sign extension happens when you cast from a signed number type consisting of a smaller number of bits to one with a larger number of bits, such as short to int for example.

> But since I'm casting from an int to a short, no sign extension should happen if I'm not mistaken. The fact that this does not issue a warning in earlier versions of Visual Studio, it leads me to believe that this must be a bug in the Visual Studio 2015 compiler (Roslyn). Am I misunderstanding how sign extension and/or the compiler works here, or is this most likely a compiler bug?

The VS2015 compiler is too eager to detect cases of sign extension in bitwise-or where CS2013 did not. The newly reported cases are such that the report does not warn about an actually dangerous coding pattern. We should adjust the implementation to be closer to VS2013's rules.

**Comment by AlekseyTs** [comment]
@gafter, Please verify.
